staging/src/k8s.io/apiserver/pkg/endpoints/filters/limit_request.go

```diff
@@ -0,0 +1,79 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package filters
+
+import (
+	"fmt"
+	"net/http"
+	"strings"
+	"sync"
+
+	"k8s.io/apiserver/pkg/authentication/user"
+	"k8s.io/apiserver/pkg/endpoints/handlers/responsewriters"
+	genericapirequest "k8s.io/apiserver/pkg/endpoints/request"
+	"k8s.io/apiserver/pkg/server/healthz"
+	"k8s.io/klog"
+)
+
+// WithLimitUntilHealthy bloxks all requests to the handler unless the API is healthy for the first time or it is a health or liveness check
+func WithLimitUntilHealthy(handler http.Handler, checks []healthz.HealthChecker) http.Handler {
+	healthyFirstTime := false
+	isSelfRequest := false
+	mu := &sync.Mutex{}
+
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		urlPaths := strings.Split(r.URL.Path, "/")
+		allowedURLS := map[string]struct{}{
+			"healthz": {},
+			"livez":   {},
+			"readyz":  {},
+		}
+
+		mu.Lock()
+		defer mu.Unlock()
```

Comment:
With this all request handlers run in serial, not good. As said before, an atomic value is much better.